### PR TITLE
Fix original purchase date when different base plans grant different entitlements

### PR DESCRIPTION
**Pull request by vegaro**
I found an edge case when purchasing two base plans from the same product, when each base plan grants a different entitlement. This is a very weird scenario but figured we should fix it.

We build the `original_purchase_date` of the entitlement using the product section of the customer info response, but that information is only valid for the entitlement of the base plan purchased. Which means if another base plan was purchased after the original expired, the information in the product section is only valid for the entitlement that base plan is granting.

To give an example. Imagine `prod_1:p1m` grants `entitlement_1` and `prod_1:p1y` grants `entitlement_2`. `p1m` is purchased, and `entitlement_1` is granted. After `p1m` expires, `p1y` is purchased and `entitlement_2` is granted. At that point, the subscription section of the customer info response will look like:

```
"subscriptions":{
   "prod_1":{
      "billing_issues_detected_at":null,
      "is_sandbox":false,
      "original_purchase_date":"2023-04-11T17:59:03.285Z",
      "purchase_date":"2023-04-11T17:59:03.285Z",
      "store":"play_store",
      "unsubscribe_detected_at":null,
      "product_plan_identifier":"p1y",
      "expires_date":"2024-04-11T17:59:03.285Z",
      "period_type":"normal"
   }
}
```

The `entitlements` section of the response looks like:

```
"entitlements":{
   "entitlement_1":{
      "expires_date":"2023-04-11T17:59:03.285Z",
      "product_identifier":"prod_1",
      "purchase_date":"2023-03-11T16:59:03.285Z",
      "product_plan_identifier":"p1m"
   },
   "entitlement_2":{
      "expires_date":"2024-04-11T17:59:03.285Z",
      "product_identifier":"prod_1",
      "purchase_date":"2023-04-11T17:59:03.285Z",
      "product_plan_identifier":"p1y"
   }
}
```

If we use the information of the `subscription` to build the `original_purchase_date` of `entitlement_1`, we will be setting the wrong `original_purchase_date`.

This can probably be fixed in the backend, but here's a proposed SDK solution, where I check if the product information matches the base plan that's granting the entitlement.

### CODE REVIEW

File: common/src/main/java/com/revenuecat/purchases/common/EntitlementInfoFactories.kt [hunk 1]
```diff
@@ -112,6 +113,40 @@ internal fun JSONObject.buildEntitlementInfo(
...
+//   "purchase_date":"2023-04-11T17:59:03.285Z",
+//   "store":"play_store",
+//   "unsubscribe_detected_at":null,
+//   "product_plan_identifier":"not_bw",
+//   "expires_date":"2023-04-12T18:59:01.115Z",
+//   "period_type":"normal"
+// }
+private fun JSONObject.calculateOriginalPurchaseDate(productData: JSONObject): Date =
+    if (productData.optNullableString(ProductResponseJsonKeys.PRODUCT_PLAN_IDENTIFIER) ==
+        this.optNullableString(EntitlementsResponseJsonKeys.PRODUCT_PLAN_IDENTIFIER)) {
```
[the remark this answers is not in the document]
Comment (tonidero) [reply common/src/main/java/com/revenuecat/purchases/common/EntitlementInfoFactories.kt:144]: Ahh right 👍

**Review by tonidero — APPROVED** [review]
I do think this logic would be better to live in the backend, but I'm ok merging this for now. Should we create a ticket to track that?

**Comment by tonidero** [comment]
Oh actually, tests are failing

**Comment by vegaro** [comment]
Fixed the test. It had some leftovers from another branch

**Comment by vegaro** [comment]
Looks like this is going to be solved in the backend https://linear.app/revenuecat/issue/CAT-919/add-original-purchase-date-to-the-entitlements-section-of-the-be so I think we can close this? What do you think @tonidero ? We can also merge this and update when the backend has shipped

**Comment by tonidero** [comment]
> Looks like this is going to be solved in the backend https://linear.app/revenuecat/issue/CAT-919/add-original-purchase-date-to-the-entitlements-section-of-the-be so I think we can close this? What do you think @tonidero ? We can also merge this and update when the backend has shipped

I guess it depends on the timeframe for this to be fixed in the backend and whether we think this might cause issues for developers. I'm not sure of the use cases for this field, as in what are developers using it for exactly? If we think this is unlikely to cause issues in the short term we can wait for the backend fix and just close this.

**Comment by vegaro** [comment]
I am going to close this, it's very unlikely two base plans from the same offer grants different entitlements so I am going to defer this to the backend fix

cc @swehner

**Pull request closed without merging**